transport/transporters/rabbitmq/transporter/transporter.go

```diff
@@ -0,0 +1,261 @@
+func (t RabbitMQTransporter) shutdown() {
+	t.log.Info("shutting down transporter")
+	t.shutdownHandler.CancelFunc() // initiate shutdown on other modules as well
+
+	if r := recover(); r != nil {
+		t.log.Warnf("Recovered in RabbitMQTransporter %s", r)
+	}
+
+	defer func() {
+		// recover if channel is already closed
+		recover()
+	}()
+
+	t.log.Debug("closing progress channel")
+	close(t.txnsWritten)
+}
+
+// StartTransporting gets a RabbitMQ connection from the ConnectionGetter and
+// passes it to transport()
+func (t RabbitMQTransporter) StartTransporting() {
+	t.log.Info("starting transporter")
+	defer t.shutdown()
+
+	for {
+		select {
+		case <-t.shutdownHandler.TerminateCtx.Done():
+			t.log.Debug("received terminateCtx cancellation")
+			return
+		default:
+		}
+
+		conn, err := t.connMan.GetConnection(t.shutdownHandler.TerminateCtx)
+		if err != nil {
+			return
+		}
+		t.transport(conn)
+	}
+}
+
+// transport reads in message batches, outputs its String to RabbitMQ and then sends a progress report on the batch
+func (t RabbitMQTransporter) transport(conn wabbit.Conn) {
+	var b interface{}
+	var ok bool
+	var err error
+
+	t.channel, err = conn.Channel()
+	if err != nil {
+		t.log.WithError(err).Fatal("Could not open channel to RabbitMQ")
+	}
+	defer func() {
+		err = t.channel.Close()
+		if err != nil {
+			t.log.WithError(err).Error("Error while closing channel")
+		}
+	}()
+	t.log.Info("Created new channel")
+	closeNotify := t.channel.NotifyClose(make(chan wabbit.Error))
+	t.publishNotify = t.channel.NotifyPublish(make(chan wabbit.Confirmation, t.batchSize))
+	err = t.channel.Confirm(false)
+	if err != nil {
+		t.log.WithError(err).Fatal("Could not turn on confirmations for channel")
+	}
+	confirms := uint64(0)
+
+	for {
+		select {
+		case <-t.shutdownHandler.TerminateCtx.Done():
+			t.log.Debug("received terminateCtx cancellation")
+			return
+		case <-closeNotify:
+			t.log.Debug("channel closed")
+			return
+
+		case b, ok = <-t.inputChan:
+			// pass
+		}
+
+		select {
+		case <-t.shutdownHandler.TerminateCtx.Done():
+			t.log.Debug("received terminateCtx cancellation")
+			return
+		case <-closeNotify:
+			return
+		default:
+			// pass
+		}
+
+		if !ok {
+			t.log.Warn("input channel is closed")
+			return
+		}
+
+		genericBatch, ok := b.(*batch.GenericBatch)
+
+		if !ok {
+			panic("Batch is not a GenericBatch")
+		}
+
+		messages := genericBatch.GetPayload()
+		messagesSlice, ok := messages.([]*marshaller.MarshalledMessage)
+
+		if !ok {
+			panic("Batch payload is not a []*marshaller.MarshalledMessage")
+		}
+
+		// Begin timer
+		start := TimeSource.UnixNano()
+
+		err = t.sendMessages(closeNotify, messagesSlice)
+		if err != nil {
+			t.log.WithError(err).Error("Could not transport messages")
+			t.statsChan <- stats.NewStatCount("rabbitmq_transport", "failure", 1, TimeSource.UnixNano())
+			return
+		}
+
+		confirms, err = t.waitForConfirmations(closeNotify, uint64(len(messagesSlice)), confirms)
```

Comment:
The transport doesn't have retry logic for attempting to publish a message or a batch of messages again. Is this something the underlying library does?

Reply:
My assumption here was that a failure in delivery would be caused by a disconnection, which is handled by the transport layer by exiting `transport()` and retrying with a new connection and a new call to `transport()` in the for-loop in `StartTransporting()`.

Reply:
Do you think additional retry logic is appropriate here?  There are other ways a call could fail, but they wouldn't get resolved by a retry, and I believe I've avoided them here.  (Turning on the mandatory flag, so that every message ends up in a queue is one example.  For my use case, it's okay if there isn't a queue, and it's easy enough for the user to create a queue for the messages anyways.)

Reply:
If there is any failure in `transport` then the batch would have already been read off of the `inputChan`. We would wind up in a scenario where there is a gap because the batch was consumed but never retried or acknowledged. This case has to be covered.

There's obviously a handful of ways to do this. I recommend the transport function calling out to a get connection function which would create a new or return an existing connection. Something like this:

```
transportWithRetry(context, batch) error {
   conn, err = getConn()
   if err... return err (this would cause retrier to retry)

  sendMessages(... batch)
  waitForConfirmations(... batch)
}

getConn() {
   if conn is closed
        return new conn, err
   else
       return conn, nil
}

StartTransporting() {
   for {
      read batch / term context

      retry(transportWithRetry)
   }
}
```

Reply:
I see what you're saying.  I'll address this.

Reply:
I've brought the rabbitmq transporter much closer to the kinesis transporter's retry handling.  I believe it is now identical, or as close as possible, in behavior.